x/pairing/keeper/epoch_payments.go

```diff
@@ -63,58 +63,94 @@ func (k Keeper) GetAllEpochPayments(ctx sdk.Context) (list []types.EpochPayments
 	}
 
+	// get this epoch's epochPayments object
 	epochPayments, found, key := k.GetEpochPaymentsFromBlock(ctx, epoch)
 	if !found {
-		epochPayments = types.EpochPayments{Index: key, ClientsPayments: []*types.ProviderPaymentStorage{userPaymentProviderStorage}}
+		// this epoch doesn't have a epochPayments object, create one with the providerPaymentStorage object from before
+		epochPayments = types.EpochPayments{Index: key, ProviderPaymentStorageKeys: []string{userPaymentProviderStorage.GetIndex()}}
 	} else {
-		epochPayments.ClientsPayments = append(epochPayments.ClientsPayments, userPaymentProviderStorage)
+		// this epoch has a epochPayments object -> make sure this payment is not already in this object
+		providerPaymentStorageKeyFound := false
+		for _, providerPaymentStorageKey := range epochPayments.GetProviderPaymentStorageKeys() {
+			if providerPaymentStorageKey == userPaymentProviderStorage.GetIndex() {
+				providerPaymentStorageKeyFound = true
+				break
+			}
+		}
+
+		// this epoch's epochPayments object doesn't contain this providerPaymentStorage key -> append the new key
+		if !providerPaymentStorageKeyFound {
+			epochPayments.ProviderPaymentStorageKeys = append(epochPayments.ProviderPaymentStorageKeys, userPaymentProviderStorage.GetIndex())
+		}
 	}
 
+	// update the epochPayments object
 	k.SetEpochPayments(ctx, epochPayments)
+
 	return usedCUProviderTotal, nil
 }
 
+// Function to remove all epochPayments objects from a specific epoch
 func (k Keeper) RemoveAllEpochPaymentsForBlock(ctx sdk.Context, blockForDelete uint64) error {
-	// remove the old epochs
+	// get the epochPayments object of blockForDelete
 	epochPayments, found, key := k.GetEpochPaymentsFromBlock(ctx, blockForDelete)
 	if !found {
 		// return fmt.Errorf("did not find any epochPayments for block %d", blockForDelete.Num)
+		// no epochPayments object -> do nothing
 		return nil
 	}
-	userPaymentsStorages := epochPayments.ClientsPayments
-	for _, userPaymentStorage := range userPaymentsStorages {
-		uniquePaymentStoragesCliPro := userPaymentStorage.UniquePaymentStorageClientProvider
-		for _, uniquePaymentStorageCliPro := range uniquePaymentStoragesCliPro {
+
+	// go over the epochPayments object's providerPaymentStorageKeys
+	userPaymentsStorageKeys := epochPayments.GetProviderPaymentStorageKeys()
+	for _, userPaymentStorageKey := range userPaymentsStorageKeys {
+		// get the providerPaymentStorage object
+		userPaymentStorage, _ := k.GetProviderPaymentStorage(ctx, userPaymentStorageKey)
+
+		// go over the providerPaymentStorage object's uniquePaymentStorageClientProviderKeys
+		uniquePaymentStoragesCliProKeys := userPaymentStorage.GetUniquePaymentStorageClientProviderKeys()
+		for _, uniquePaymentStorageKey := range uniquePaymentStoragesCliProKeys {
+			// get the uniquePaymentStorageClientProvider object
+			uniquePaymentStorage, _ := k.GetUniquePaymentStorageClientProvider(ctx, uniquePaymentStorageKey)
```

Comment:
use the found and check it it is true, if not continue

Reply:
done